File: Test files/Test cases/classes.swift
//
// Copyright 2018 Vinicius Jorge Vendramini
//
// Licensed under the Hippocratic License, Version 2.1;
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// https://firstdonoharm.dev/version/2/1/license.md
//
// To the full extent allowed by law, this software comes "AS IS,"
// WITHOUT ANY WARRANTY, EXPRESS OR IMPLIED, and licensor and any other
// contributor shall not be liable to anyone for any damages or other
// liability arising from, out of, or in connection with the sotfware
// or this license, under any kind of legal claim.
// See the License for the specific language governing permissions and
// limitations under the License.
//

// gryphon insert: import kotlin.system.exitProcess

class Box {
	var a: Int = 5

	var b: String {
		get {
			return "get b"
		}
		set {
			print("set b")
		}
	}

	var c: Int?

	var d: Int? {
		get {
			return 0
		}
	}

	var e: Int? {
		return 0
	}

	func returnFive() -> Int {
		return a
	}

	func returnInt(a: Int) -> Int {
		return a
	}
}

let box1 = Box()

print(box1.a)
print(box1.returnFive())
print(box1.returnInt(a: 10))
print(box1.b)
box1.b = "whatever"
print(box1.c)

// Removes inner types: it should be `B()` and not `A.B()`
class A {
	class B {
	}

	let b = B()
}

// Test static methods
class C {
	var x = 0

	static func a() -> C {
		return C()
	}

	func b() -> C {
		return C()
	}

	static func c() -> C? {
		return C()
	}

	static func d() -> C? {
		return nil
	}

	func e() -> Int {
		print("Hello, world!")
		return 1
	}

	static func f() -> Int {
		return 1
	}
}

struct D {
	let x: Int

	// gryphon ignore
	init(x: Int) {
		self.x = x
	}

	init?(string: String?) {
		guard let string = string else {
			return nil
		}

		switch string {
		case "A": self = D(x: 0)
		case "B": self = D(x: 0)
		case "C": self = D(x: 0)
		case "D": self = D(x: 0)
		case "E": self = D(x: 0)
		default: return nil
		}
	}
}

// Test open/final
// gryphon annotation: open
class E {
	var a = 0
}

// gryphon annotation: final
class F {
	var a = 0
}

print(C.a().x)
print(C().b().x)

var a = C.c()
if let ac = a {
	print(ac.x)
}

a = C.d()
if let ad = a {
	print(ad.x)
}

print(C().e())
print(C.f())

print(D(x: 10))
print(D(string: "not supported"))
print(D(string: "A")!)

class SingleExpressionMembers {
	var one: Int {
		10
	}

	var two: Int {
		get {
			20
		}
	}

	var noSingleExpression: Int {
		fatalError("foo")
	}

	var dddddddddddddddddddddddddeeeeeeeeeee: Int {
		aaaaaaaaaaaaa(bbbbbbbbbbbbbbbb: 0, ccccccccccccc : 1, ddddddddddddddddddddddddd: 2, eeeeeeeeeee: 3)
	}

	private func aaaaaaaaaaaaa(bbbbbbbbbbbbbbbb: Int, ccccccccccccc: Int, ddddddddddddddddddddddddd: Int, eeeeeeeeeee: Int) -> Int {
		bbbbbbbbbbbbbbbb * ccccccccccccc * ddddddddddddddddddddddddd * eeeeeeeeeee
	}
}

let singleExpressionMembers = SingleExpressionMembers()
print(singleExpressionMembers.one)
print(singleExpressionMembers.two)
print(singleExpressionMembers.dddddddddddddddddddddddddeeeeeeeeeee)

// Super calls
class Superclass {
	var x = 0

	init(x: Int) {
		self.x = x
	}
}

class Subclass: Superclass {
	init(y: Int) {
		super.init(x: y + 1)
	}
}

// Test `class` annotations
class G {
	init() { }
	class func setUp() { }
}

// Test removing `override` annotations
class H: G {
	override init() { }
	override static func setUp() { }
}

File: Test files/Test cases/classes.kt
//
// Copyright 2018 Vinicius Jorge Vendramini
//
// Licensed under the Hippocratic License, Version 2.1;
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// https://firstdonoharm.dev/version/2/1/license.md
//
// To the full extent allowed by law, this software comes "AS IS,"
// WITHOUT ANY WARRANTY, EXPRESS OR IMPLIED, and licensor and any other
// contributor shall not be liable to anyone for any damages or other
// liability arising from, out of, or in connection with the sotfware
// or this license, under any kind of legal claim.
// See the License for the specific language governing permissions and
// limitations under the License.
//
import kotlin.system.exitProcess

internal open class Box {
	open var a: Int = 5
	open var b: String
		get() {
			return "get b"
		}
		set(newValue) {
			println("set b")
		}
	open var c: Int? = null
	open val d: Int?
		get() {
			return 0
		}
	open val e: Int?
		get() {
			return 0
		}

	open fun returnFive(): Int {
		return a
	}

	open fun returnInt(a: Int): Int {
		return a
	}
}

internal open class A {
	open class B {
	}

	val b: B = B()
}

internal open class C {
	companion object {
		fun a(): C {
			return C()
		}

		fun c(): C? {
			return C()
		}

		fun d(): C? {
			return null
		}

		fun f(): Int {
			return 1
		}
	}

	open var x: Int = 0

	open fun b(): C {
		return C()
	}

	open fun e(): Int {
		println("Hello, world!")
		return 1
	}
}

internal data class D(
	val x: Int
) {
	companion object {
		operator fun invoke(string: String?): D? {
			string ?: return null
			return when (string) {
				"A" -> D(0)
				"B" -> D(0)
				"C" -> D(0)
				"D" -> D(0)
				"E" -> D(0)
				else -> null
			}
		}
	}
}

internal open class E {
	open var a: Int = 0
}

internal class F {
	open var a: Int = 0
}

internal open class SingleExpressionMembers {
	open val one: Int
		get() = 10
	open val two: Int
		get() = 20
	open val noSingleExpression: Int
		get() {
			println("Fatal error: ${"foo"}"); exitProcess(-1)
		}
	open val dddddddddddddddddddddddddeeeeeeeeeee: Int
		get() = aaaaaaaaaaaaa(
			bbbbbbbbbbbbbbbb = 0,
			ccccccccccccc = 1,
			ddddddddddddddddddddddddd = 2,
			eeeeeeeeeee = 3)

	private fun aaaaaaaaaaaaa(
		bbbbbbbbbbbbbbbb: Int,
		ccccccccccccc: Int,
		ddddddddddddddddddddddddd: Int,
		eeeeeeeeeee: Int)
		: Int
		= bbbbbbbbbbbbbbbb * ccccccccccccc * ddddddddddddddddddddddddd * eeeeeeeeeee
}

internal open class Superclass {
	open var x: Int = 0

	constructor(x: Int) {
		this.x = x
	}
}

internal open class Subclass: Superclass {
	constructor(y: Int): super(x = y + 1) {
	}
}

internal open class G {
	companion object {
		fun setUp() {
		}
	}

	constructor() {
	}
}

internal open class H: G {
	companion object {
		fun setUp() {
		}
	}

	constructor() {
	}
}

fun main(args: Array<String>) {
	val box1: Box = Box()

	println(box1.a)
	println(box1.returnFive())
	println(box1.returnInt(a = 10))
	println(box1.b)

	box1.b = "whatever"

	println(box1.c)
	println(C.a().x)
	println(C().b().x)

	var a: C? = C.c()
	val ac: C? = a

	if (ac != null) {
		println(ac.x)
	}

	a = C.d()

	val ad: C? = a

	if (ad != null) {
		println(ad.x)
	}

	println(C().e())
	println(C.f())
	println(D(10))
	println(D(string = "not supported"))
	println(D(string = "A")!!)

	val singleExpressionMembers: SingleExpressionMembers = SingleExpressionMembers()

	println(singleExpressionMembers.one)
	println(singleExpressionMembers.two)
	println(singleExpressionMembers.dddddddddddddddddddddddddeeeeeeeeeee)
}
